org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/ExecutionPage.java

```diff
@@ -150,14 +166,113 @@ public void controlResized(ControlEvent e) {
+        this.progressListener = new ExecutionProgressListener(this, this.processDescription.getJob());
         this.operation.addProgressListener(this.progressListener);
 
         // return the tree as the outermost page control
         return this.filteredTree;
     }
 
+    public void onProgress(ProgressEvent progressEvent) {
+        OperationDescriptor descriptor = progressEvent.getDescriptor();
+        if (isExcluded(descriptor)) {
+            return;
+        }
+        OperationItem operationItem = this.allItems.get(descriptor);
+        if (null == operationItem) {
+            operationItem = new OperationItem((StartEvent) progressEvent);
+            this.allItems.put(descriptor, operationItem);
+            this.activeItems.put(operationItem, Boolean.FALSE);
+        } else {
+            operationItem.setFinishEvent((FinishEvent) progressEvent);
+            this.activeItems.put(operationItem, Boolean.TRUE);
+            if (isJvmTestSuite(descriptor) && operationItem.getChildren().isEmpty()) {
+                // do not display test suite nodes that have no children (unwanted artifacts from Gradle)
+                OperationItem parentOperationItem = this.allItems.get(findFirstNonExcludedParent(descriptor));
+                parentOperationItem.removeChild(operationItem);
+                return;
+            }
+        }
+
+        // attach to (first non-excluded) parent, if this is a new operation (in case of StartEvent)
+        OperationItem parentExecutionItem = this.allItems.get(findFirstNonExcludedParent(descriptor));
+        parentExecutionItem.addChild(operationItem);
+    }
+
+    private boolean isExcluded(OperationDescriptor descriptor) {
+        // ignore the 'artificial' events issued for the root test event and for each forked test
+        // process event
+        if (descriptor instanceof JvmTestOperationDescriptor) {
+            JvmTestOperationDescriptor jvmTestOperationDescriptor = (JvmTestOperationDescriptor) descriptor;
+            return jvmTestOperationDescriptor.getSuiteName() != null && jvmTestOperationDescriptor.getClassName() == null;
+        } else {
+            return false;
+        }
+    }
+
+    private OperationDescriptor findFirstNonExcludedParent(OperationDescriptor descriptor) {
+        while (isExcluded(descriptor.getParent())) {
+            descriptor = descriptor.getParent();
+        }
+        return descriptor.getParent();
+    }
+
+    public void refreshChangedItems() {
+        TreeViewer viewer = this.filteredTree.getViewer();
+        for (OperationItem item : this.activeItems.keySet()) {
+            viewer.update(item, null);
+            if (shouldBeVisible(item)) {
+                viewer.expandToLevel(item, 0);
+            }
+        }
+        viewer.refresh(false);
+
+        Iterator<OperationItem> items = this.activeItems.keySet().iterator();
```

Comment:
I think this would be simpler if we just had a second set of `removedItems` and called `clear` on that set here.